test/lightning/failure_alert_test.exs

```diff
@@ -176,77 +175,77 @@ defmodule Lightning.FailureAlertTest do
 
-      {:ok, attempt_run} =
-        AttemptRun.new()
-        |> Ecto.Changeset.put_assoc(
-          :attempt,
-          Attempt.changeset(%Attempt{}, %{
-            work_order_id: work_order.id,
-            reason_id: reason.id
-          })
-        )
-        |> Ecto.Changeset.put_assoc(
-          :run,
-          Run.changeset(%Run{}, %{
-            project_id: project.id,
-            job_id: job.id,
-            input_dataclip_id: dataclip.id
-          })
-        )
-        |> Repo.insert()
+      assert_email_sent(subject: "\"workflow-a\" failed.")
 
-      Pipeline.process(attempt_run)
+      FailureAlerter.alert_on_failure(run_3)
 
       refute_email_sent(subject: "\"workflow-a\" failed.")
     end
 
     test "does not increment the rate-limiter counter when an email is not delivered.",
-         %{attempt_run: attempt_run, work_order: work_order} do
+         %{runs: [run, _, _], workorders: [workorder, _, _]} do
       [time_scale: time_scale, rate_limit: rate_limit] =
         Application.fetch_env!(:lightning, Lightning.FailureAlerter)
 
-      Pipeline.process(attempt_run)
+      FailureAlerter.alert_on_failure(run)
 
       {:ok, {0, ^rate_limit, _, _, _}} =
-        Hammer.inspect_bucket(work_order.workflow_id, time_scale, rate_limit)
+        Hammer.inspect_bucket(workorder.workflow_id, time_scale, rate_limit)
 
       assert_email_sent(subject: "\"workflow-a\" failed.")
 
-      stub(Lightning.FailureEmail, :deliver_failure_email, fn _, _ ->
+      Mimic.stub(Lightning.FailureEmail, :deliver_failure_email, fn _, _ ->
         {:error}
       end)
 
-      Pipeline.process(attempt_run)
+      FailureAlerter.alert_on_failure(run)
 
       refute_email_sent(subject: "\"workflow-a\" failed.")
 
       # nothing changed
       {:ok, {0, ^rate_limit, _, _, _}} =
-        Hammer.inspect_bucket(work_order.workflow_id, time_scale, rate_limit)
+        Hammer.inspect_bucket(workorder.workflow_id, time_scale, rate_limit)
+    end
+
+    test "failure alert is sent on run complete", %{
+      attempts: [attempt, _, _]
+    } do
+      Lightning.Stub.reset_time()
+
+      {:ok, bearer, _} =
+        Workers.Token.generate_and_sign(
+          %{},
+          Lightning.Config.worker_token_signer()
+        )
+
+      {:ok, %{}, socket} =
+        LightningWeb.WorkerSocket
+        |> socket("socket_id", %{token: bearer})
+        |> subscribe_and_join(
+          LightningWeb.AttemptChannel,
+          "attempt:#{attempt.id}",
+          %{"token" => Workers.generate_attempt_token(attempt)}
+        )
+
+      %{runs: [run]} = Repo.preload(attempt, :runs)
+
+      push(socket, "run:complete", %{
+        "run_id" => run.id,
+        "output_dataclip_id" => Ecto.UUID.generate(),
+        "output_dataclip" => ~s({"foo": "bar"}),
+        "reason" => "normal"
+      })
+
+      # Wait a bit to make sure run:complete is handled
+      # TODO: Check with Taylor DOWNS and team if this is good enough
```

Comment:
@taylordowns2000 please in the AM, would you mind checking this out ?

Reply:
looks good. reason "normal" should be reason "failed", but otherwise it's great.